runtime-modules/referendum/src/lib.rs

```diff
@@ -0,0 +1,740 @@
+// TODO: module documentation
+// TODO: adjust all extrinsic weights
+
+// NOTE: This module is instantiable pallet as described here https://substrate.dev/recipes/3-entrees/instantiable.html
+// No default instance is provided.
+
+/////////////////// Configuration //////////////////////////////////////////////
+#![allow(clippy::type_complexity)]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+// used dependencies
+use codec::{Codec, Decode, Encode};
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage, error::BadOrigin, traits::Get, Parameter,
+    StorageValue,
+};
+use sp_arithmetic::traits::{BaseArithmetic, One};
+use sp_runtime::traits::{MaybeSerialize, Member};
+use std::marker::PhantomData;
+use system::ensure_signed;
+
+use std::collections::HashSet;
+
+// declared modules
+mod mock;
+mod tests;
+
+/////////////////// Data Structures ////////////////////////////////////////////
+
+#[derive(Encode, Decode, PartialEq, Eq, Debug)]
+pub enum ReferendumStage {
+    Void,
+    Voting,
+    Revealing,
+}
+
+impl Default for ReferendumStage {
+    fn default() -> ReferendumStage {
+        ReferendumStage::Void
+    }
+}
+
+#[derive(Encode, Decode, PartialEq, Eq, Debug, Default)]
+pub struct SealedVote<Hash, CurrencyBalance> {
+    commitment: Hash,
+    stake: CurrencyBalance,
+}
+
+#[derive(Clone, Encode, Decode, PartialEq, Eq, Debug)]
+pub enum ReferendumResult<ReferendumOption, VotePower> {
+    Winners(Vec<(ReferendumOption, VotePower)>),
+    ExtraWinners(Vec<(ReferendumOption, VotePower)>),
+    NotEnoughWinners(Vec<(ReferendumOption, VotePower)>),
+    NoVotesRevealed,
+}
+
+impl<T, U> Default for ReferendumResult<T, U> {
+    fn default() -> ReferendumResult<T, U> {
+        ReferendumResult::NoVotesRevealed
+    }
+}
+
+/////////////////// Trait, Storage, Errors, and Events /////////////////////////
+
+pub trait Trait<I: Instance>: system::Trait {
+    /// The overarching event type.
+    type Event: From<Event<Self, I>> + Into<<Self as system::Trait>::Event>;
+
+    /// Maximum number of options in one referendum.
+    type MaxReferendumOptions: Get<u64>;
+    type ReferendumOption: Parameter
+        + Member
+        + BaseArithmetic
+        + Codec
+        + Default
+        + Copy
+        + MaybeSerialize
+        + PartialEq
+        + From<u64>
+        + Into<u64>;
+
+    /// Currency balance used for stakes.
+    type CurrencyBalance: Parameter
+        + Member
+        + BaseArithmetic
+        + Codec
+        + Default
+        + Copy
+        + MaybeSerialize
+        + PartialEq
+        + From<u64>
+        + Into<u64>;
+
+    type VotePower: Parameter
```

Comment:
VotePower gives more flexibility to how votes should be counted. Depending on `fn caclulate_vote_power(account_id, stake)` implementation it can handle for example these situations:
- each account has one vote (`return 1;`)
- the voting power is linear with a stake (`return stake;`) - council's case
- 1st class citizen has higher vote power than 2nd class citizen (`return if account_id == MY_1ST_CLASS_CITIZEN {stake * 10} else {stake} `)
- possibly others